### PR TITLE
Bump starknet dep, adjust wait_for_tx, use new devnet for tests

### PR DESCRIPTION
**Pull request by THenry14**
Closes https://github.com/foundry-rs/starknet-foundry/issues/546
Closes https://github.com/foundry-rs/starknet-foundry/issues/372
Closes https://github.com/foundry-rs/starknet-foundry/issues/811

## Introduced changes

- bumped starknet to newest version
- adjusted wait_for_tx to new rpc schemas
- replaced pythonic devnet with devnet-rs for running tests

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - diagnostic paths referring to `tests` folder
 
+### Cast
+
+#### Changed
+- bump starknet-rs dependency
+- replaced starknet-devnet with starknet-devnet-rs for tests
```
Comment (cptartur) [inline CHANGELOG.md:20]: Not sure if necessary in changelog

File: crates/cast/src/lib.rs [hunk 2]
```diff
@@ -187,7 +183,13 @@ fn get_account_from_keystore<'a>(
             .ok_or_else(|| anyhow::anyhow!("Failed to get address from account JSON file - make sure the account is deployed"))?
     )?;
 
-    Ok(SingleOwnerAccount::new(provider, signer, address, chain_id))
+    Ok(SingleOwnerAccount::new(
+        provider,
+        signer,
+        address,
+        chain_id,
+        ExecutionEncoding::Legacy,
```
Comment (cptartur) [inline crates/cast/src/lib.rs:191]: Why `::Legacy`?

File: crates/cast/src/lib.rs [hunk 3]
```diff
@@ -238,56 +246,33 @@ pub async fn wait_for_tx(
     retries: u8,
 ) -> Result<&str> {
     println!("Transaction hash: {tx_hash:#x}");
-
-    let mut maybe_receipt: Option<MaybePendingTransactionReceipt> = None;
     for i in (1..retries).rev() {
```
Comment (cptartur) [inline crates/cast/src/lib.rs:249]: Didn't know that the famous `wait_for_tx` from starknet.py made it's apperance here in cast 😅

File: crates/cast/tests/e2e/deploy.rs [hunk 4]
```diff
@@ -94,7 +94,7 @@ async fn test_contract_not_declared() {
     let snapbox = runner(&args);
     let output = String::from_utf8(snapbox.assert().get_output().stderr.clone()).unwrap();
 
-    assert!(output.contains("Class with hash 0x1 is not declared."));
+    assert!(output.contains("Transaction execution has failed."));
```
Comment (cptartur) [inline crates/cast/tests/e2e/deploy.rs:97]: Isn't it a regression?

File: crates/cast/tests/e2e/invoke.rs [hunk 5]
```diff
@@ -49,11 +49,9 @@ async fn test_contract_does_not_exist() {
     ]);
 
     let snapbox = runner(&args);
+    let output = String::from_utf8(snapbox.assert().success().get_output().stderr.clone()).unwrap();
 
-    snapbox.assert().stderr_matches(indoc! {r#"
-        command: invoke
-        error: Contract not found
-    "#});
+    assert!(output.contains("Transaction execution has failed."));
```
Comment (cptartur) [inline crates/cast/tests/e2e/invoke.rs:54]: Same here

File: CHANGELOG.md (hunk 1 above)
Comment (THenry14) [reply CHANGELOG.md:20]: A few people asked about it, so I figured I should include it. If you disagree, I can remove it no problem.

File: crates/cast/src/lib.rs (hunk 2 above)
Comment (THenry14) [reply crates/cast/src/lib.rs:191]: We'll do migration to cairo1 accounts in separate task (as this has to be adjusted in devnet as well for proper testing)

File: crates/cast/src/lib.rs (hunk 3 above)
Comment (THenry14) [reply crates/cast/src/lib.rs:249]: Curtesy of Marcin :D

File: crates/cast/tests/e2e/deploy.rs (hunk 4 above)
Comment (THenry14) [reply crates/cast/tests/e2e/deploy.rs:97]: This is devnet side; I did not check recently, but real nodes did return different messages (and we are not dependent on the message body, only on tx status). Will try to find some time to run cast against integration to check what are the messages, and raise an issue to devnet if there are differences. For now I don't think this is blocking in any way.

File: crates/cast/src/lib.rs [hunk 6]
```diff
@@ -238,56 +246,33 @@ pub async fn wait_for_tx(
...
-                Invoke(receipt) => receipt.status,
-                Declare(receipt) => receipt.status,
-                Deploy(receipt) => receipt.status,
-                DeployAccount(receipt) => receipt.status,
-                L1Handler(receipt) => receipt.status,
-            }
-        } else {
-            println!("Received transaction. Status: Pending");
+        if i > 1 {
             sleep(Duration::from_secs(5));
```
Comment (war-in) [inline crates/cast/src/lib.rs:269]: I think I don't understand this `if`. Why does sleep happen only for `i > 1`?

File: CHANGELOG.md (hunk 1 above)
Comment (cptartur) [reply CHANGELOG.md:20]: We're generating releases' changelogs from this file so for me it makes sense to keep it focused on user-facing changes. Though I don't have a very strong opinion on that - if you think it's useful to include this change here we can do it.

Though I'd reword it so it's more obvious that you mean our own tests and not something related with testing, so it won't get confused with forge things.

File: crates/cast/tests/e2e/deploy.rs (hunk 4 above)
Comment (cptartur) [reply crates/cast/tests/e2e/deploy.rs:97]: Doesn't have to be resolved in this PR, but I'd create an issue to check this later.

File: crates/cast/src/lib.rs (hunk 6 above)
Comment (THenry14) [reply crates/cast/src/lib.rs]: Ahhhh yes I am big dzban and forgot to remove this `if`, I used it for troubleshooting during testing phase. Sorry for that, and good catch :+1:.

File: crates/cast/tests/e2e/deploy.rs (hunk 4 above)
Comment (THenry14) [reply crates/cast/tests/e2e/deploy.rs:97]: Sure, https://github.com/foundry-rs/starknet-foundry/issues/1000

File: CHANGELOG.md (hunk 1 above)
Comment (THenry14) [reply CHANGELOG.md]: Ok makes sense actually. I have an opened comms channel with the people interested in this, so I can give them heads up this way, no reason to pollute the changelog. Deleted.

File: CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - diagnostic paths referring to `tests` folder
 
+### Cast
+
+#### Changed
+
```
Comment (cptartur) [inline CHANGELOG.md:19]:
```suggestion
```

**Review by cptartur — APPROVED** [review]
